source/common/stats/statsd.cc

```diff
@@ -33,44 +33,56 @@ Writer::~Writer() {
   }
 }
 
-void Writer::writeCounter(const std::string& name, uint64_t increment) {
-  std::string message(fmt::format("envoy.{}:{}|c", name, increment));
-  send(message);
-}
-
-void Writer::writeGauge(const std::string& name, uint64_t value) {
-  std::string message(fmt::format("envoy.{}:{}|g", name, value));
-  send(message);
-}
-
-void Writer::writeTimer(const std::string& name, const std::chrono::milliseconds& ms) {
-  std::string message(fmt::format("envoy.{}:{}|ms", name, ms.count()));
-  send(message);
-}
-
-void Writer::send(const std::string& message) {
+void Writer::write(const std::string& message) {
   ::send(fd_, message.c_str(), message.size(), MSG_DONTWAIT);
 }
 
 UdpStatsdSink::UdpStatsdSink(ThreadLocal::SlotAllocator& tls,
-                             Network::Address::InstanceConstSharedPtr address)
-    : tls_(tls.allocateSlot()), server_address_(address) {
+                             Network::Address::InstanceConstSharedPtr address, const bool use_tag)
+    : tls_(tls.allocateSlot()), server_address_(address), use_tag_(use_tag) {
   tls_->set([this](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
     return std::make_shared<Writer>(this->server_address_);
   });
 }
 
 void UdpStatsdSink::flushCounter(const Counter& counter, uint64_t delta) {
-  tls_->getTyped<Writer>().writeCounter(counter.name(), delta);
+  std::string message(
+      fmt::format("envoy.{}:{}|c{}", getName(counter), delta, buildTagStr(counter.tags())));
+  tls_->getTyped<Writer>().write(message);
 }
 
 void UdpStatsdSink::flushGauge(const Gauge& gauge, uint64_t value) {
-  tls_->getTyped<Writer>().writeGauge(gauge.name(), value);
+  std::string message(
+      fmt::format("envoy.{}:{}|g{}", getName(gauge), value, buildTagStr(gauge.tags())));
+  tls_->getTyped<Writer>().write(message);
 }
 
 void UdpStatsdSink::onHistogramComplete(const Histogram& histogram, uint64_t value) {
   // For statsd histograms are all timers.
-  tls_->getTyped<Writer>().writeTimer(histogram.name(), std::chrono::milliseconds(value));
+  std::string message(fmt::format("envoy.{}:{}|ms{}", getName(histogram),
+                                  std::chrono::milliseconds(value).count(),
+                                  buildTagStr(histogram.tags())));
+  tls_->getTyped<Writer>().write(message);
+}
+
+std::string UdpStatsdSink::getName(const Metric& metric) {
+  if (use_tag_) {
+    return metric.tagExtractedName();
+  } else {
+    return metric.name();
+  }
+}
+
+const std::string UdpStatsdSink::buildTagStr(const std::vector<Tag>& tags) {
+  if (!use_tag_ || tags.empty()) {
+    return {};
```

Comment:
nit: I would probably use `""` here as that is more idiomatic.

Reply:
Agreed: 010936280390f12fb249f2eb291d048178aa5f98